File: 39주차/단어변환.py
from collections import deque
def check(a,b):# 두 문자열에서 다른 문자가 1개만 있는지 확인.
    cnt =0 
    for i in range(len(a)):
        if a[i]!=b[i]:
            cnt += 1
    if cnt == 1:
        return True
    return False
def solution(begin, target, words):
    
    q = deque()
    q.append((begin, 0))
    visited = [0]*len(words)
    while q: # bfs로 가장 먼저 target에 도달할 때 cnt를 반환.
        now,cnt = q.popleft()
        if now == target:
            return cnt
        for i,v in enumerate(words):
            if not visited[i]:# 이미 방문했던 문자열이라면 이후에도 같은 결과를 내기에 방문처리하기.
                if check(v,now):
                    visited[i] = 1
                    q.append((v,cnt+1))# 변환 한 번 했으므로 +1.
    return 0 # 변환을 할 수 없다면 0 반환.
    
# 프로그래머스 bfs/dfs 단어변환 복습.

File: 47주차/단어변환.js
function solution(begin, target, words) {
  var answer = 50;
  //변환이 안된다면 0 반환.
  let visited = {};
  for (let w of words) visited[w] = 0;
  function dfs(L, now) {
    if (now === target) {
      answer = Math.min(answer, L);
      return;
    }
    for (let w of words) {
      let cnt = 0;
      for (let i = 0; i < w.length; i++) {
        if (w[i] !== now[i]) cnt += 1;
      }
      if (cnt === 1 && visited[w] === 0) {
        visited[w] = 1;
        dfs(L + 1, w);
        visited[w] = 0;
      }
    }
  }
  dfs(0, begin);
  return answer === 50 ? 0 : answer;
}
// 한글자만 다른 방향으로 뻗어나가기. 최소 거칠 수 있는 단계 구하기
// 프로그래머스 dfs
